planner/core/common_plans.go

```diff
@@ -297,9 +303,32 @@ func (e *Execute) OptimizePreparedPlan(ctx context.Context, sctx sessionctx.Cont
 	}
 	e.SnapshotTS = snapshotTS
 	e.Stmt = prepared.Stmt
+	e.handleStalenessOption(sctx)
 	return nil
 }
 
+func (e *Execute) handleStalenessOption(sctx sessionctx.Context) {
+	vars := sctx.GetSessionVars()
+	readTS := vars.TxnReadTS.PeakTxnReadTS()
+	// If readTS > 0, it means we are handling the following case:
+	// 1. set transaction read only as of timestamp ts
+	// 2. execute prepared statement
+	if readTS > 0 {
```

Comment:
I am not sure if these two cases should be handled. I mean you bypass all checks in `Preprocessor` and assume `select * from t` is well-defined as of time `ts`, e.g. `t as of ts` may not exist.

I think we can reset the `ts`, `staleness`, `txnScope` using the result of:

```
ret := &PreprocessorReturn{InfoSchema: is}
err := Preprocess(sctx, prepared.Stmt, InPrepare, WithPreprocessorReturn(ret))
```

It is better than dirty hack.